### PR TITLE
Matrix permanent

### PR DESCRIPTION
**Pull request by jrexmo**
#### References to other Issues or PRs
#19035

#### Brief description of what is fixed or changed
Matrix now includes permanent. Same as #19035, excluding the dumb removals that occurred earlier - should resolve build issues.

#### Release Notes

* matrices
  * Added a new method for computing the permanent of a matrix

### CODE REVIEW

File: sympy/matrices/determinant.py [hunk 1]
```diff
@@ -828,3 +828,43 @@ def _minor_submatrix(M, i, j):
...
+
+    >>> from sympy import Matrix
+    >>> M = Matrix([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
+    >>> M.permanent()
+    450
+
+    References
+    ==========
+
+    1. Prof. Frank Ben's notes: https://math.berkeley.edu/~bernd/ban275.pdf
```
Comment (czgdp1807) [inline sympy/matrices/determinant.py:848]:
```suggestion
    .. [1] Prof. Frank Ben's notes: https://math.berkeley.edu/~bernd/ban275.pdf
```

File: sympy/matrices/determinant.py [hunk 2]
```diff
@@ -828,3 +828,43 @@ def _minor_submatrix(M, i, j):
...
+    if not M.is_square:
+        raise NonSquareMatrixError()
+
+    n = M.rows
+
+    perm_list = list(permutations(range(n)))
+
+    total = 0
+    # Computing permanent by summing over all permuatations on (0, ... n-1)
+    # TODO: find a faster way to do this, maybe a sparse method, maybe Gaussian Elim
```
Comment (oscarbenjamin) [inline sympy/matrices/determinant.py:862]: Apparently Gaussian elimination can not be used for the permanent but there is a more efficient formula here:
https://en.wikipedia.org/wiki/Computing_the_permanent#Ryser_formula

File: sympy/matrices/determinant.py [hunk 3]
```diff
@@ -828,3 +828,43 @@ def _minor_submatrix(M, i, j):
     cols = [a for a in range(M.cols) if a != j]
 
     return M.extract(rows, cols)
+
+
+def _permanent(M):
+    """Returns the permanent of a matrix as defined in [1]
+
```
Comment (asmeurer) [inline sympy/matrices/determinant.py:835]: It would be helpful to add the definition here as well.

File: sympy/matrices/determinant.py (hunk 1 above)
Comment (asmeurer) [reply sympy/matrices/determinant.py:848]: Also include a link to the Wikipedia article.

**Comment by czgdp1807** [comment]
@jrexmo Would you like to complete this PR? Thanks for the contributions.

**Comment by jrexmo** [comment]
Yes, I will be finishing up shortly - have been having some time-consuming system issues.

File: sympy/matrices/matrices.py [hunk 4]
```diff
@@ -982,12 +984,10 @@ def _handle_creation_inputs(cls, *args, **kwargs):
             # Matrix(MatrixSymbol('X', 2, 2))
             elif isinstance(args[0], Basic) and args[0].is_Matrix:
                 return args[0].rows, args[0].cols, args[0].as_explicit()._mat
-
-            elif isinstance(args[0], mp.matrix):
-                M = args[0]
-                flat_list = [cls._sympify(x) for x in M]
+            elif isinstance(args[0], mp.matrix):
```
Comment (sylee957) [inline sympy/matrices/matrices.py:987]:
```suggestion
            elif isinstance(args[0], mp.matrix):
```

Comment (sylee957) [reply sympy/matrices/matrices.py:987]: But anyway are these changes needed?

File: sympy/matrices/expressions/factorizations.py [hunk 5]
```diff
@@ -30,6 +30,10 @@ class SofSVD(Factorization):
 class VofSVD(Factorization):
     predicates = Q.orthogonal,
 
+class PofPolar(Factorization):
+    predicates = Q.nonnegative,
```
Comment (sylee957) [inline sympy/matrices/expressions/factorizations.py:34]: I think that you should clean up the changes for polar decompositions in #18958

File: sympy/matrices/determinant.py [hunk 6]
```diff
@@ -826,3 +826,58 @@ def _minor_submatrix(M, i, j):
...
+    if not M.is_square:
+        raise NonSquareMatrixError()
+
+    n = M.rows
+    s = list(range(n))
+
+    subsets = []
+    for i in range(n):
+        subsets += list(map(list, itertools.combinations(s, i + 1)))
```
Comment (sylee957) [inline sympy/matrices/determinant.py:866]: These lines look redundant.

**Comment by naveensaigit** [comment]
I looked into Ryser's algorithm and got an implementation in Julia [here](https://numbersandshapes.net/post/permanents_and_rysers_algorithm/)

I coded a sample and it looks like this -
``` python
from sympy import prod, zeros, ones
from timeit import default_timer as timer 

def perm(M):
    n = M.shape[0]
    gd = zeros(2**n-1)
    gd[0] = 1

    lut = {2**i : (i,1) for i in range(n)}
    lut.update({-2**i : (i,-1) for i in range(n)})
    r = M[0, :]
    s = -1
    pm = s*prod(r)
    for i in range(2,2**n):
        if i%2==0:
    	    gd[i-1] = 2*gd[i//2-1]
        else:
    	    gd[i-1] = (-1)**(i//2)
        r += M[lut[gd[i-1]][0],:]*lut[gd[i-1]][1]
        s *= -1
        pm += s*prod(r)
    return(pm * (-1)**n)

if __name__=="__main__":  
    start = timer() 
    print(perm(ones((11,11))))
    print('Time Elapsed : ',timer()-start)

Output -
39916800 # Permanent of n x n ones is equal to n!
Time Elapsed : 20.8072631
```

The downside is that the same algorithm is more than 10x faster in Julia!

**Comment by naveensaigit** [comment]
This could be closed now

**Pull request closed without merging**